### PR TITLE
Stop cleanUpNbsps orphaning CKEditor's zero-width filling char in titles (BL-16808)

### PR DESCRIPTION
**Pull request by andrew-polk**
**Problem.** Follow-up to #8305. A tester found that on 6.4 a title typed as "One Two" with Enter (or Shift+Enter) pressed at the start of "Two" still reaches the Publish tab and bloomlibrary.org as "OneTwo", even though #8305 now puts a line break between the two lines. Show Hidden Characters reveals a zero-width space at the start of the second line. 6.5 does not have the problem.

**Cause.** CKEditor plants a zero-width "filling char" (U+200B) at the caret whenever it re-selects a caret that sits next to an inline element, such as the line-break span Shift+Enter inserts, and removes it later by holding a reference to that text node. Bloom's half-second tidy-up after every keystroke took a CKEditor bookmark and restored it, which is such a re-select, and it also assigned the box's `innerHTML` whether or not anything had changed. Each rewrite detached the node CKEditor was tracking, so the character stayed as ordinary text, a fresh one was planted on the next pause, and the result was saved. Chromium renders a newline adjacent to U+200B as nothing at all, so the line break became invisible. Reproduced in a running 6.4 build over CDP, and master fixed both halves in BL-16558 and BL-16717, which is why 6.5 is clean.

**What the PR does.** Ports the two master changes: `cleanUpNbsps` assigns `innerHTML` only when it actually converted a non-breaking space, and the tidy-up takes a bookmark only when a tool is active or the box holds a comment or an nbsp, the only cases in which anything rewrites it. With both, the tester's Enter and Shift+Enter sequences are clean at every step in a running Bloom, Show Hidden Characters included, and the saved title is "One \nTwo". Five new tests check node identity and the gate. Not done: cleaning books that already carry a stray zero-width space, and stripping one that CKEditor itself plants after, say, a Backspace next to a line break and that Show Hidden Characters then orphans; master strips those at save (BL-16490), which was judged too risky for 6.4 because Khmer, Thai and Myanmar text uses U+200B legitimately.

Ref: https://issues.bloomlibrary.org/youtrack/issue/BL-16808

[Devin review](https://devinreview.com/BloomBooks/BloomDesktop/pull/8332)

- - -
This change is [Reviewable](https://reviewable.io/reviews/BloomBooks/BloomDesktop/8332)

### CODE REVIEW

File: src/BloomBrowserUI/bookEdit/toolbox/toolbox.ts [hunk 1]
```diff
@@ -1675,9 +1684,10 @@ export function cleanUpNbsps(editableDiv: HTMLElement) {
                 editableDivText.substring(0, j) +
                 " " +
                 editableDivText.substring(j + 1);
+            replacedAnNbsp = true;
```
Comment (andrew-polk) [inline src/BloomBrowserUI/bookEdit/toolbox/toolbox.ts:1713]: [Devin] **Bug**: Converted spaces still corrupt titles

`cleanUpNbsps` still assigns `editableDiv.innerHTML` whenever it converts a non-breaking space. If CKEditor has an active filling-char text node anywhere in that box at that moment, the rebuild detaches the node while keeping its U+200B in the serialized HTML. CKEditor later cleans the detached node rather than the replacement, so the zero-width space stays in the saved content and can hide the title's line break. (Devin rated this non-severe.)

Comment (andrew-polk) [reply src/BloomBrowserUI/bookEdit/toolbox/toolbox.ts:1713]: [Claude Fable 5.1] Deliberately left as is. This is the same limit master has carried since BL-16558: the rewrite still happens on the (rare) keystroke pause where an nbsp actually needs converting, and on that pause a filling char in the box is orphaned. Closing that window fully needs either rebuilding only the converted text nodes or preserving CKEditor's filling-char tracking across the rebuild, both of which are a larger change than a 6.4 release branch should take. Master additionally strips stray U+200B at save time (BL-16490), which we decided not to backport because it also removes legitimate zero-width spaces in Khmer, Thai and Myanmar text. The residual case is called out in the test ideas on BL-16808.

File: src/BloomBrowserUI/bookEdit/toolbox/toolboxSpec.ts [hunk 2]
```diff
@@ -124,4 +124,56 @@ describe("toolbox tests", () => {
...
+        expect(div.innerHTML).toBe("<p>A b&nbsp;</p>");
+        expect(div.querySelector("p")!.firstChild).toBe(textNodeBefore);
+    });
+
+    it("cleanUpNbsps does rebuild the content when there is something to convert", () => {
+        const div = document.createElement("div");
+        div.innerHTML = "<p>A&nbsp;b c</p>";
+        const textNodeBefore = div.querySelector("p")!.firstChild;
+        // Sanity check the setup: this nbsp is one we expect to be converted.
+        expect(textNodeBefore?.textContent).toBe("A\u00A0b c");
```
Comment (andrew-polk) [inline src/BloomBrowserUI/bookEdit/toolbox/toolboxSpec.ts:180]: [Devin] **Investigate**: Regression test misses combined trigger

The tests separate node preservation from conversion. None combines a converted space with CKEditor's active U+200B marker, leaving the reported edge case uncovered.

Comment (andrew-polk) [reply src/BloomBrowserUI/bookEdit/toolbox/toolboxSpec.ts:180]: [Claude Fable 5.1] Not adding one: the combined case is a known, accepted limit of this fix (see the Bug thread on toolbox.ts), so a test for it would have to assert the behavior we consider undesirable, which pins the limit in place rather than guarding a fix. The two node-identity tests and the bookmark test cover what this PR changes.

**Comment by andrew-polk** [comment]
[Claude Fable 5.1] Consulted Devin on 2026-09-08 20:27 UTC up to commit `3037a63ac8cbd7f7cfb4710a69395e110c53c22a`. It raised one non-severe bug and one Investigate flag, both about the same accepted limit (the box is still rebuilt on the rare pause where a non-breaking space is converted); each has its own resolved thread above with the reasoning. No informational items. CI passed.

**Comment by andrew-polk** [comment]
[Claude Fable 5.1] Consulted Devin on 2026-09-08 21:11 UTC up to commit `32b5e5e168d186fff43dc0da72277472ee520a6c` (and on the two commits before it this run). Each re-review repeated the same non-severe bug and Investigate flag about the accepted limit already answered in the resolved threads above (the box is still rebuilt on the rare pause where a non-breaking space is converted); nothing new. No informational items. CI passed.

**Comment by andrew-polk** [comment]
[Claude Fable 5.1] Consulted Devin on 2026-09-08 21:48 UTC up to commit `cd867a2bfd3edd62bf5a5a3f9f07748bf91b624c` (comment-wording change). Same two repeated findings as before, already answered in the resolved threads above; nothing new.

**Review by StephenMcConnel — APPROVED** [review]
> @StephenMcConnel reviewed 2 files and all commit messages.
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomDesktop/8332)* status: all files reviewed, 2 unresolved discussions.

**Review by StephenMcConnel — APPROVED** [review]
> @StephenMcConnel resolved 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomDesktop/8332)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on andrew-polk).
